### PR TITLE
[DPM-160][DPM-146] make first game action in new_game request and some refacto…

### PR DESCRIPTION
**Pull request by algys**
…ring

### CODE REVIEW

File: eventprocessor/handlers.go [hunk 1]
```diff
@@ -1,61 +1,145 @@
...
-    "github.com/eoscanada/eos-go"
-
-    "github.com/rs/zerolog/log"
+	"context"
+	"encoding/json"
+	eventlistener "github.com/DaoCasino/platform-action-monitor-client"
+	"github.com/eoscanada/eos-go"
+	"github.com/rs/zerolog/log"
+	"platform-backend/models"
+    "time"
```
Comment (justefg) [inline eventprocessor/handlers.go:10]: go fmt

File: eventprocessor/handlers.go [hunk 2]
```diff
@@ -1,61 +1,145 @@
...
-}
+    err := p.repos.GameSession.AddGameSessionUpdate(ctx, &models.GameSessionUpdate{
+        SessionID:  session.ID,
+        UpdateType: models.SessionStartedUpdate,
+        Timestamp:  time.Now(),
+        Data:       event.Data,
+    })
+    if err !=  nil {
+    	return err
+	}
```
Comment (justefg) [inline eventprocessor/handlers.go:24]: go fmt

File: eventprocessor/processor.go [hunk 3]
```diff
@@ -2,32 +2,38 @@ package eventprocessor
...
 // game events
 const (
-	gameStarted eventlistener.EventType = iota
-	actionRequest
-	signidicePartOneRequest
-	signidicePartTwoRequest
-	gameFinished
-	gameFailed
-	gameMessage
+	gameStarted             = 0
```
Comment (justefg) [inline eventprocessor/processor.go:15]: iota ?

File: eventprocessor/processor.go [hunk 4]
```diff
@@ -2,32 +2,38 @@ package eventprocessor
...
+	signidicePartOneRequest: onSignidicePartOneRequest,
+	gameFinished:            onGameFinished,
+	gameFailed:              onGameFailed,
+	gameMessage:             onGameMessage,
+}
 
 type EventProcessor struct {
 	repos      *repositories.Repos
-	blockChain *blockchain.Blockchain
+	blockchain *blockchain.Blockchain
```
Comment (justefg) [inline eventprocessor/processor.go:36]: camelCase

File: blockchain/blockchain.go [hunk 5]
```diff
@@ -40,6 +46,9 @@ type Blockchain struct {
 	ChainId             eos.Checksum256
 	PlatformAccountName string
 	sponsorUrl          string
+	optsMutex           sync.Mutex
```
Comment (justefg) [inline blockchain/blockchain.go:49]: why did you add mutex?

File: eventprocessor/handlers.go [hunk 6]
```diff
@@ -1,61 +1,145 @@
...
-    err = p.useCases.Signidice.PerformSignidice(ctx, game.Contract, data.Digest, gs.BlockchainSesID)
-    if err != nil {
-        return err
+	// first action already processed by back-end
+    if session.State == models.GameStartedInBC {
+        return nil
     }
 
-    log.Debug().Msgf("Successfully signed and sent signidice for session: %d", gs.ID)
+    err := p.repos.GameSession.AddGameSessionUpdate(ctx, &models.GameSessionUpdate{
```
Comment (justefg) [inline eventprocessor/handlers.go:42]: i think we should rename the method to `AddSessionUpdate` - `GameSession` already implies it is a a game session update
Similar with EventProcessor

File: game_sessions/repository/localstorage/game_session_update.go [hunk 7]
```diff
@@ -14,9 +14,7 @@ func (r *GameSessionsLocalRepo) GetGameSessionUpdates(ctx context.Context, id ui
 	sessionUpdates := make([]*models.GameSessionUpdate, 0)
 	ses := r.gameSessions[id]
 	if ses.Updates != nil {
-		for _, v := range ses.Updates {
-			sessionUpdates = append(sessionUpdates, v)
-		}
+		sessionUpdates = append(sessionUpdates, ses.Updates...)
```
Comment (justefg) [inline game_sessions/repository/localstorage/game_session_update.go:17]: why not just `return ses.Updates, nil`

File: server/api/interfaces/error.go [hunk 8]
```diff
@@ -1,6 +1,45 @@
 package interfaces
```
Comment (justefg) [inline server/api/interfaces/error.go:1]: why `interfaces` package. I don't see any interfaces here

File: server/api/api.go [hunk 9]
```diff
@@ -100,26 +121,44 @@ func (api *WsApi) ProcessRawRequest(context context.Context, messageType int, me
...
+		// process request
+		wsResp, handlerError := handler.handler(context, &interfaces.ApiRequest{
 			UseCases: api.useCases,
 			Repos:    api.repos,
 			User:     user,
 			Data:     &messageObj,
 		})
+
+		if handlerError != nil {
+			if handlerError.Code == interfaces.InternalError {
```
Comment (justefg) [inline server/api/api.go:150]: maybe implement `String()` method for `HandlerError` ?

File: server/api/interfaces/error.go [hunk 10]
```diff
@@ -1,6 +1,45 @@
...
-	Code    uint64  `json:"code"`
-	Message string `json:"message"`
+	Code    WsErrorCode `json:"code"`
+	Message string      `json:"message"`
+}
+
+type HandlerError struct {
+	Code          WsErrorCode
+	InternalError error
+}
```
Comment (justefg) [inline server/api/interfaces/error.go:13]: i recommend implementing `Error() string` that way you'll implement `error` interface

File: server/api/interfaces/error.go (hunk 8 above)
Comment (algys) [reply server/api/interfaces/error.go:1]: basically, `interfaces` contain WS API interfaces, but yes "interfaces" not in terms of golang

File: eventprocessor/handlers.go (hunk 1 above)
Comment (algys) [reply eventprocessor/handlers.go]: fixed

File: blockchain/blockchain.go (hunk 5 above)
Comment (algys) [reply blockchain/blockchain.go:49]: because GetTxOpts can be invoked from different threads.

File: eventprocessor/handlers.go (hunk 6 above)
Comment (algys) [reply eventprocessor/handlers.go]: rename AddSessionUpdate to Update ?

File: server/api/api.go (hunk 9 above)
Comment (algys) [reply server/api/api.go:150]: why? what profit?

File: eventprocessor/handlers.go (hunk 2 above)
Comment (algys) [reply eventprocessor/handlers.go:24]: fixed

File: eventprocessor/processor.go (hunk 3 above)
Comment (algys) [reply eventprocessor/processor.go:15]: I don't like iota in cases like that. Original events numeration controlled by game contract. Contract doesn't  guaranties that order is without gaps and also doesn't guaranties that new event will have "next" number as  type.

File: game_sessions/repository/localstorage/game_session_update.go (hunk 7 above)
Comment (algys) [reply game_sessions/repository/localstorage/game_session_update.go:17]: fixed

File: eventprocessor/processor.go (hunk 3 above)
Comment (justefg) [reply eventprocessor/processor.go:15]: oh, I see

File: eventprocessor/handlers.go (hunk 6 above)
Comment (justefg) [reply eventprocessor/handlers.go]: just bringing attention, no need to rename in this PR

File: server/api/api.go (hunk 9 above)
Comment (justefg) [reply server/api/api.go:150]: you could just write `handlerError.Error()` instead of a large `if else block`

File: server/api/interfaces/error.go (hunk 8 above)
Comment (justefg) [reply server/api/interfaces/error.go:1]: then please rename it, it's confusing - I expected to see a bunch of interfaces, although in go we create interfaces where they're used

https://github.com/golang/go/wiki/CodeReviewComments#interfaces

File: server/api/api.go (hunk 9 above)
Comment (algys) [reply server/api/api.go:150]: hmm, unfortunately,  It won't help me(
Anyway, I need if else block to handle different errors type, in case of internal I log with error level, in another case use just debug one

File: server/api/interfaces/error.go (hunk 8 above)
Comment (algys) [reply server/api/interfaces/error.go]: renamed
